From: Steven Rostedt <srostedt@redhat.com>
Subject: rwlock: remove waiter from pi_list

If the pending owner was a reader and we have multiple readers than
we need to remove it from the pi list.

[ Thomas Gleixner added the grabing of the pi_locks to the removal ]

Signed-off-by: Steven Rostedt <srostedt@redhat.com>
---
 kernel/rtmutex.c |   13 +++++++++++++
 1 file changed, 13 insertions(+)

Index: linux-2.6.24.7-rt27/kernel/rtmutex.c
===================================================================
--- linux-2.6.24.7-rt27.orig/kernel/rtmutex.c	2009-02-08 00:04:49.000000000 -0500
+++ linux-2.6.24.7-rt27/kernel/rtmutex.c	2009-02-08 00:04:49.000000000 -0500
@@ -1685,6 +1685,7 @@ rt_read_slowunlock(struct rw_mutex *rwm,
 {
 	struct rt_mutex *mutex = &rwm->mutex;
 	struct rt_mutex_waiter *waiter;
+	struct task_struct *pendowner;
 	struct reader_lock_struct *rls;
 	unsigned long flags;
 	unsigned int reader_count;
@@ -1793,6 +1794,7 @@ rt_read_slowunlock(struct rw_mutex *rwm,
 			rwm->owner = RT_RW_PENDING_READ;
 	}
 
+	pendowner = waiter->task;
 	wakeup_next_waiter(mutex, savestate);
 
 	/*
@@ -1808,6 +1810,17 @@ rt_read_slowunlock(struct rw_mutex *rwm,
 	if (rt_mutex_has_waiters(mutex)) {
 		waiter = rt_mutex_top_waiter(mutex);
 		rwm->prio = waiter->task->prio;
+		/*
+		 * If readers still own this lock, then we need
+		 * to update the pi_list too. Readers have a separate
+		 * path in the PI chain.
+		 */
+		if (reader_count) {
+			spin_lock(&pendowner->pi_lock);
+			plist_del(&waiter->pi_list_entry,
+				  &pendowner->pi_waiters);
+			spin_unlock(&pendowner->pi_lock);
+		}
 	} else
 		rwm->prio = MAX_PRIO;
 
